src/coral_models/asr/ngram_model.py

```diff
@@ -1,81 +1,78 @@
 
 import os
+import subprocess
+import tempfile
 from pathlib import Path
-from shutil import rmtree
 
 from datasets import Dataset, load_dataset
 from huggingface_hub import Repository
+from omegaconf import DictConfig
 from pyctcdecode.decoder import build_ctcdecoder
 from transformers import AutoProcessor, Wav2Vec2ProcessorWithLM
 
 
-def train_ngram_model(
-    model_id: str,
-    dataset_id: str = "DDSC/reddit-da-asr-preprocessed",
-    split: str = "train",
-    n: int = 5,
-) -> None:
+def train_ngram_model(cfg: DictConfig) -> None:
     """Trains an ngram language model.
 
     Args:
-        model_id (str):
-            The model id of the finetuned speech model, which we will merge
-            with the ngram model.
-        dataset_id (str, optional):
-            The dataset to use for training. Defaults to
-            'DDSC/reddit-da-asr-preprocessed'.
-        split (str, optional):
-            The split to use for training. Defaults to 'train'.
-        n (int, optional):
-            The ngram order to use for training. Defaults to 5.
+        cfg (DictConfig):
+            Hydra configuration dictionary.
     """
-    # Ensure that the data folder exists
-    data_dir = Path("data")
-    if not data_dir.exists():
-        data_dir.mkdir()
-
-    # Ensure that the models folder exists
-    models_dir = Path("models")
-    if not models_dir.exists():
-        models_dir.mkdir()
-
     # Load the dataset
-    try:
-        dataset = load_dataset(dataset_id, split=split, use_auth_token=True)
-    except ValueError:
-        dataset = Dataset.from_file(f"{dataset_id}/dataset.arrow")
+    dataset = load_dataset(
+        path=cfg.model.decoder.dataset_id,
+        name=cfg.model.decoder.dataset_subset,
+        split=cfg.model.decoder.dataset_split,
+        use_auth_token=os.getenv("HUGGINGFACE_HUB_TOKEN"),
+    )
+    assert isinstance(dataset, Dataset)
 
-    # Dump dataset to a text file
-    text_path = data_dir / "text_data.txt"
-    with open(text_path, "w") as f:
-        f.write(" ".join(dataset["text"]))
+    # Dump dataset to a temporary text file
+    text_file = tempfile.NamedTemporaryFile(mode="w", delete=False, suffix=".txt")
+    text_file.write(" ".join(dataset["text"]))
 
     # Ensure that the `kenlm` directory exists, and download if otherwise
-    kenlm_dir = Path("kenlm")
+    kenlm_dir = Path.home() / ".cache" / "kenlm"
     if not kenlm_dir.exists():
-        os.system("wget -O - https://kheafield.com/code/kenlm.tar.gz | tar xz")
+        kenlm_url = "https://kheafield.com/code/kenlm.tar.gz"
+        subprocess.Popen(["wget", "-O", "-", kenlm_url], stdout=subprocess.PIPE)
+        subprocess.Popen(["tar", "-xz"], stdin=subprocess.PIPE, stdout=subprocess.PIPE)
+        subprocess.Popen(
+            ["rm", "-rf", "kenlm.tar.gz"],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+        )
+        subprocess.Popen(["mv", "kenlm", str(kenlm_dir)], stdin=subprocess.PIPE)
```

Comment:
I would suggest just doing this in Python instead of calling subprocesses.

```
import urllib.request
import tarfile

def download_and_extract(url, target_dir):
    # Download the file
    with urllib.request.urlopen(url) as response:
        data = response.read()

    # Extract the file
    with tarfile.open(fileobj=io.BytesIO(data)) as tar:
        tar.extractall(path=target_dir)
```

Then you can just `download_and_extract(kenlm_url, kenlm_dir)`.

Reply:
Good catch @lfwa! I've included `wait()` calls now to ensure that they are run sequentially.

Reply:
@rlrs Yeah that's more readable for sure, I agree.